### DISCUSSION TITLE
Best practice for feeding into WorkflowHub

### DISCUSSION THREAD

**Discussion by AlasdairGray** [BioSchemas/specifications#581] [General]
> > Hopefully they can then be used to feed into WorkflowHub.

> What's the ideal way for this to work, we'll probably be submitting workflows to workflowhub ourselves.
> 
> - Do we annotate our WFs with ld+json, knowing WFHub will do it too, resulting in "two copies" for anyone scanning for workflows?
> - Do we not annotate ours, knowing the "canonical" version can be kept in WF Hub?
> - Do we just annotate ours, and somehow WF Hub will discover these and pick them up, saving us from doing anything?
> 
> I keep looking into working with wf hub but still haven't found good options for syncing all of our workflows programmatically, and I'm very interested in anything that helps us automate that!

The above text originated in a discussion on the web site pull request https://github.com/BioSchemas/bioschemas.github.io/pull/547#issuecomment-1150776857

**Comment by fbacall** [comment]
Hi, I just got alerted to this.

I know @simleo is connected to several activities around this, in particular I'm thinking about Galaxy IWC workflows and the LifeMonitor GitHub app. Is there a solution that @hexylena could explore?

**Reply by CaroleGoble** [reply]
we do need general advice for this - the idea is that ANY workflow system can markup using bioschemas and we can consume it in the Hub. Otherwise we are stuck with non-scalable one-to-one specialist solutions. One for galaxy, one for nextflow, one for snakemake etc etc

**Reply by hexylena** [reply]
> Is there a solution that @hexylena could explore?

I wish I had bandwidth for this @fbacall! Agreed with @CaroleGoble I'd love to see some general markup letting us completely skip the (expensive) implementation of a customer submitter to WFH.

**Comment by stain** [comment]
Discussed in WorkflowHub call today - further work needed to formalize this as with just a landing page and without the RO-Crate or repository it's much harder to "get" the workflow consistently. 

for IWC there is also the issue on what to index then -- https://github.com/galaxyproject/iwc is a GitHub repository without any Bioschemas markup exposed on the Web -- they could potentially be on http://galaxyproject.github.io/iwc ? 

It is stated the IWC workflows are listed on the public instances, but pages like https://usegalaxy.eu/workflows/list_published has a flat list of all the workflows, not a single page per workflow as traditional for Bioschema markup.  Bioschemas have had a few ways on how to deal with these collections but no clear guidance on the Bioschemas website.

It would be unclear what WorkflowHub would link to, perhaps <https://usegalaxy.org/u/iwc/w/dctmd-calculations-with-gromacs-20> (just picking one of the instances?) as landing page that can have the markup and then links to the .ga file https://usegalaxy.org/workflow/export_to_file?id=195a608c481040de (or ideally Workflow RO-Crate)

These are the kind of details and conventions that would have to be sorted before standalone Bioschemas could be a reliable import mechanism for WorkflowHub. In the first instance we therefore went with RO-Crate and CWL as they can be maintained in Git repositories and imported from there directly as done at the moment.

**Reply by hexylena** [reply]
> These are the kind of details and conventions that would have to be sorted before standalone Bioschemas could be a reliable import mechanism for WorkflowHub

I guess it's the same as the training materials case though right? We have index pages that list multiple tutorials, but they're not marked up, the tutorials are. Instead the sitemap.xml is scanned, all pages visited, and when markup is present, it's extracted and used.

But agreed, nice to have all of this worked out.

**Comment by simleo** [comment]
Currently, the submission of IWC workflows to WorkflowHub and LifeMonitor works as follows:

* A [Python script](https://github.com/galaxyproject/iwc/blob/23c9c04c9cab23052dcffd0de58a4c848a1cac90/workflows/gen_crates.py) that I wrote is run automatically by GitHub Actions when a PR that adds / updates a workflow is merged, right before the workflow repository is deployed to [iwc-workflows](https://github.com/iwc-workflows). The script generates [Workflow Testing RO-Crate](https://crs4.github.io/life_monitor/workflow_testing_ro_crate) metadata, effectively turning the repository into an RO-Crate that's compatible with both WorkflowHub and LifeMonitor
* A scraper developed by @fbacall runs daily, importing the RO-Crates thus generated into WorkflowHub. Since the workflows are already in RO-Crate form, the metadata that get into WorkflowHub are those found in the RO-Crate
* I manually register all new IWC workflows on WorkflowHub to LifeMonitor. Avoiding this manual step is one of the reasons behind the development of the LifeMonitor GitHub app

The plan is to replace this soon, having the LifeMonitor GitHub app handle all or part of the above steps. Both with the old and the new process, however, the registration process acts on the workflow repositories, not on a web page for the workflow that might have bioschemas markup embedded in it. So what would be nice to have is a way to add Bioschemas / Schema.org metadata to the source repository. This would add up to the metadata that's already being harvested, e.g., the "creator", "license", "release" etc. fields in the `.ga` file.

[CWL has a formalism](https://www.commonwl.org/user_guide/topics/metadata-and-authorship.html) to add EDAM types to parameters and arbitrary metadata (Schema.org or other) to the workflow. Does / could Galaxy suppor something like that? If not, a convention could be set up to allow workflow authors to specify additional metadata through one or more additional files, e.g. in JSON-LD format.

Note that a good chunk of the metadata, i.e., the part related to the workflow's internals (steps, tools, parameters and their relationships), can often be extracted automatically from the workflow. In [Workflow Run RO-Crate](https://www.researchobject.org/workflow-run-crate) we are defining a profile called [Provenance Run Crate](https://www.researchobject.org/workflow-run-crate/profiles/provenance_run_crate) that can represent such metadata (*prospective provenance*), together with information about an execution of the workflow (*retrospective provenance*). The prospective provenance part is independent from specific executions, so it could be lifted up to [Workflow RO-Crate](https://about.workflowhub.eu/Workflow-RO-Crate/) and be supported by WorkflowHub (e.g., for diagrams, tabular descriptions, indexing by tool, ...). In [runcrate](https://github.com/ResearchObject/runcrate) (Workflow Run RO-Crate's toolkit) we're extracting this description from CWL workflows, and @pauldg is working on something similar for Galaxy workflows.

**Reply by hexylena** [reply]
> The plan is to replace this soon, having the LifeMonitor GitHub app handle all or part of the above steps

(as one of the galaxy training leads) best news I've heard all year. Sounds like it would simplify a *lot* for us.
